File: lib_image_editor/src/main/java/technivance/github/io/rohyme/editimage/fragment/crop/CropFragment.java
package technivance.github.io.rohyme.editimage.fragment.crop;

import android.graphics.Bitmap;
import android.graphics.Typeface;
import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.View.OnClickListener;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.ColorRes;
import androidx.core.content.ContextCompat;

import com.theartofdev.edmodo.cropper.CropImageView;

import org.jetbrains.annotations.NotNull;

import technivance.github.io.rohyme.R;
import technivance.github.io.rohyme.editimage.EditImageActivity;
import technivance.github.io.rohyme.editimage.ModuleConfig;
import technivance.github.io.rohyme.editimage.fragment.BaseEditFragment;
import technivance.github.io.rohyme.editimage.interfaces.OnLoadingDialogListener;
import technivance.github.io.rohyme.editimage.view.imagezoom.ImageViewTouchBase;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.schedulers.Schedulers;


public class CropFragment extends BaseEditFragment {
    public static final int INDEX = ModuleConfig.INDEX_CROP;
    public static final String TAG = CropFragment.class.getName();

    private static int SELECTED_COLOR = R.color.white;
    private static int UNSELECTED_COLOR = R.color.text_color_gray_3;

    private LinearLayout ratioList;
    private CropImageView cropPanel;
    private OnLoadingDialogListener loadingDialogListener;

    private CropRationClick cropRatioClick = new CropRationClick();
    private TextView selectedTextView;

    private CompositeDisposable disposables = new CompositeDisposable();

    public static CropFragment newInstance() {
        return new CropFragment();
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    @Override
    public View onCreateView(@NotNull LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_edit_image_crop, null);
    }

    private void setUpRatioList() {
        ratioList.removeAllViews();
        LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.WRAP_CONTENT,
                LinearLayout.LayoutParams.WRAP_CONTENT);

        params.gravity = Gravity.CENTER;
        params.leftMargin = 20;
        params.rightMargin = 20;

        RatioText[] ratioTextList = RatioText.values();
        for (int i = 0; i < ratioTextList.length; i++) {
            TextView text = new TextView(ensureEditActivity());
            toggleButtonStatus(text, false);
            text.setTextSize(15);
            text.setAllCaps(true);
            text.setTypeface(text.getTypeface(), Typeface.BOLD);
            text.setText(getResources().getText(ratioTextList[i].getRatioTextId()));
            ratioList.addView(text, params);

            if (i == 0) {
                selectedTextView = text;
            }

            text.setTag(ratioTextList[i]);
            text.setOnClickListener(cropRatioClick);
        }
        toggleButtonStatus(selectedTextView, true);
    }


    private final class CropRationClick implements OnClickListener {
        @Override
        public void onClick(View view) {
            toggleButtonStatus(selectedTextView, false);

            TextView currentTextView = (TextView) view;
            toggleButtonStatus(currentTextView, true);

            selectedTextView = currentTextView;

            RatioText ratioText = (RatioText) currentTextView.getTag();
            if (ratioText == RatioText.FREE) {
                cropPanel.setFixedAspectRatio(false);
            } else if (ratioText == RatioText.FIT_IMAGE) {
                Bitmap currentBmp = ensureEditActivity().getMainBit();
                cropPanel.setAspectRatio(currentBmp.getWidth(), currentBmp.getHeight());
            } else {
                AspectRatio aspectRatio = ratioText.getAspectRatio();
                cropPanel.setAspectRatio(aspectRatio.getAspectX(), aspectRatio.getAspectY());
            }
        }
    }

    private void toggleButtonStatus(TextView view, boolean isActive) {
        view.setTextColor(getColorFromRes((isActive) ? SELECTED_COLOR : UNSELECTED_COLOR));
        view.setTypeface((isActive) ? Typeface.DEFAULT_BOLD : Typeface.DEFAULT);
    }

    private int getColorFromRes(@ColorRes int resId) {
        return ContextCompat.getColor(ensureEditActivity()  , resId);
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        loadingDialogListener = ensureEditActivity();

        View backToMenu = getView().findViewById(R.id.back_to_main);
        ratioList = getView().findViewById(R.id.ratio_list_group);
        setUpRatioList();
        this.cropPanel = ensureEditActivity().cropPanel;
        backToMenu.setOnClickListener(new BackToMenuClick());
    }

    @Override
    public void onShow() {
        ensureEditActivity().mode = EditImageActivity.MODE_CROP;

        ensureEditActivity().mainImage.setVisibility(View.GONE);
        cropPanel.setVisibility(View.VISIBLE);
        ensureEditActivity().mainImage.setImageBitmap(ensureEditActivity().getMainBit());
        ensureEditActivity().mainImage.setDisplayType(ImageViewTouchBase.DisplayType.FIT_TO_SCREEN);
        ensureEditActivity().mainImage.setScaleEnabled(false);

        ensureEditActivity().bannerFlipper.showNext();
        cropPanel.setImageBitmap(ensureEditActivity().getMainBit());
        cropPanel.setFixedAspectRatio(false);
    }


    private final class BackToMenuClick implements OnClickListener {
        @Override
        public void onClick(View v) {
            backToMain();
        }
    }


    @Override
    public void backToMain() {
        ensureEditActivity().mode = EditImageActivity.MODE_NONE;
        cropPanel.setVisibility(View.GONE);
        ensureEditActivity().mainImage.setVisibility(View.VISIBLE);

        ensureEditActivity().mainImage.setScaleEnabled(true);
        ensureEditActivity().bottomGallery.setCurrentItem(0);

        if (selectedTextView != null) {
            selectedTextView.setTextColor(getColorFromRes(UNSELECTED_COLOR));
        }

        ensureEditActivity().bannerFlipper.showPrevious();
    }


    public void applyCropImage() {
        disposables.add(getCroppedBitmap()
                .subscribeOn(Schedulers.computation())
                .observeOn(AndroidSchedulers.mainThread())
                .doOnSubscribe(subscriber -> loadingDialogListener.showLoadingDialog())
                .doFinally(() -> loadingDialogListener.dismissLoadingDialog())
                .subscribe(bitmap -> {
                    ensureEditActivity().changeMainBitmap(bitmap, true);
                    backToMain();
                }, e -> {
                    e.printStackTrace();
                    backToMain();
                    Toast.makeText(getContext(), "Error while saving image", Toast.LENGTH_SHORT).show();
                }));
    }

    private Single<Bitmap> getCroppedBitmap() {
        return Single.fromCallable(() -> cropPanel.getCroppedImage());
    }

    @Override
    public void onStop() {
        disposables.clear();
        super.onStop();
    }

    @Override
    public void onDestroy() {
        disposables.dispose();
        super.onDestroy();
    }
}

File: lib_image_editor/src/main/java/iamutkarshtiwari/github/io/ananas/editimage/fragment/crop/CropFragment.kt
package iamutkarshtiwari.github.io.ananas.editimage.fragment.crop

import android.graphics.Bitmap
import android.graphics.Typeface
import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.view.View.OnClickListener
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.annotation.ColorRes
import androidx.core.content.ContextCompat
import com.theartofdev.edmodo.cropper.CropImageView
import iamutkarshtiwari.github.io.ananas.R
import iamutkarshtiwari.github.io.ananas.editimage.EditImageActivity
import iamutkarshtiwari.github.io.ananas.editimage.ModuleConfig
import iamutkarshtiwari.github.io.ananas.editimage.fragment.BaseEditFragment
import iamutkarshtiwari.github.io.ananas.editimage.view.imagezoom.ImageViewTouchBase
import io.reactivex.Single
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.schedulers.Schedulers
import kotlinx.android.synthetic.main.fragment_edit_image_crop.*


class CropFragment : BaseEditFragment() {

    private var cropPanel: CropImageView? = null

    private var selectedTextView: TextView? = null

    private val disposables = CompositeDisposable()

    private val croppedBitmap: Single<Bitmap>
        get() = Single.fromCallable { cropPanel?.croppedImage }


    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_edit_image_crop, null)
    }

    override fun onDestroyView() {
        selectedTextView =null
        cropPanel = null
        super.onDestroyView()
    }

    private fun setUpRatioList() {
        ratio_list_group.removeAllViews()
        val params = LinearLayout.LayoutParams(
            LinearLayout.LayoutParams.WRAP_CONTENT,
            LinearLayout.LayoutParams.WRAP_CONTENT
        )

        params.gravity = Gravity.CENTER
        params.leftMargin = 20
        params.rightMargin = 20

        val ratioTextList = RatioText.values()
        for (i in ratioTextList.indices) {
            val text = TextView(activity)
            toggleButtonStatus(text, false)
            text.textSize = 15f
            text.isAllCaps = true
            text.setTypeface(text.typeface, Typeface.BOLD)
            text.text = resources.getText(ratioTextList[i].ratioTextId)
            ratio_list_group.addView(text, params)

            if (i == 0) {
                selectedTextView = text
            }

            text.tag = ratioTextList[i]
            text.setOnClickListener {
                toggleButtonStatus(selectedTextView!!, false)

                val currentTextView = it as TextView
                toggleButtonStatus(currentTextView, true)

                selectedTextView = currentTextView

                val ratioText = currentTextView.tag as RatioText
                when {
                    ratioText === RatioText.FREE -> cropPanel?.setFixedAspectRatio(false)
                    ratioText === RatioText.FIT_IMAGE -> {
                        val currentBmp = ensureEditActivity().mainBit
                        cropPanel?.setAspectRatio(currentBmp.width, currentBmp.height)
                    }
                    else -> {
                        val aspectRatio = ratioText.aspectRatio
                        cropPanel?.setAspectRatio(aspectRatio.aspectX, aspectRatio.aspectY)
                    }
                }
            }
        }
        toggleButtonStatus(selectedTextView!!, true)
    }


    private fun toggleButtonStatus(view: TextView, isActive: Boolean) {
        view.setTextColor(getColorFromRes(if (isActive) SELECTED_COLOR else UNSELECTED_COLOR))
        view.typeface = if (isActive) Typeface.DEFAULT_BOLD else Typeface.DEFAULT
    }

    private fun getColorFromRes(@ColorRes resId: Int): Int {
        return ContextCompat.getColor(activity, resId)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)


        setUpRatioList()
        this.cropPanel = ensureEditActivity().cropPanel
        back_to_main.setOnClickListener(BackToMenuClick())
    }

    override fun onShow() {
        activity.mode = EditImageActivity.MODE_CROP

        activity.mainImage.visibility = View.GONE
        cropPanel?.visibility = View.VISIBLE
        activity.mainImage.setImageBitmap(activity.mainBit)
        activity.mainImage.displayType = ImageViewTouchBase.DisplayType.FIT_TO_SCREEN
        activity.mainImage.setScaleEnabled(false)

        activity.bannerFlipper.showNext()
        cropPanel?.setImageBitmap(activity.mainBit)
        cropPanel?.setFixedAspectRatio(false)
    }


    private inner class BackToMenuClick : OnClickListener {
        override fun onClick(v: View) {
            backToMain()
        }
    }


    override fun backToMain() {
        activity.mode = EditImageActivity.MODE_NONE
        cropPanel?.visibility = View.GONE
        activity.mainImage.visibility = View.VISIBLE

        activity.mainImage.setScaleEnabled(true)
        activity.bottomGallery.currentItem = 0

        if (selectedTextView != null) {
            selectedTextView!!.setTextColor(getColorFromRes(UNSELECTED_COLOR))
        }

        activity.bannerFlipper.showPrevious()
    }


    fun applyCropImage() {
        disposables.add(croppedBitmap
            .subscribeOn(Schedulers.computation())
            .observeOn(AndroidSchedulers.mainThread())
            .doOnSubscribe { subscriber -> ensureEditActivity()?.showLoadingDialog() }
            .doFinally { ensureEditActivity()?.dismissLoadingDialog() }
            .subscribe({ bitmap ->
                activity.changeMainBitmap(bitmap, true)
                backToMain()
            }, { e ->
                e.printStackTrace()
                backToMain()
                Toast.makeText(context, "Error while saving image", Toast.LENGTH_SHORT).show()
            })
        )
    }

    override fun onStop() {
        disposables.clear()
        super.onStop()
    }

    override fun onDestroy() {
        disposables.dispose()
        super.onDestroy()
    }

    companion object {
        const val INDEX = ModuleConfig.INDEX_CROP
        val TAG = CropFragment::class.java.name

        private val SELECTED_COLOR = R.color.white
        private val UNSELECTED_COLOR = R.color.text_color_gray_3

        fun newInstance(): CropFragment {
            return CropFragment()
        }
    }
}
